File: ios/Shared/Sync/FxACrypto.swift
import CryptoKit
import Foundation

enum FxACrypto {
    static let namespace = "identity.mozilla.com/picl/v1/"

    static func hex(_ data: Data) -> String {
        data.map { String(format: "%02x", $0) }.joined()
    }

    static func unhex(_ string: String) throws -> Data {
        let raw = string.trimmingCharacters(in: .whitespacesAndNewlines)
        guard raw.count.isMultiple(of: 2) else { throw SyncError.crypto("invalid hex") }
        var data = Data()
        data.reserveCapacity(raw.count / 2)
        var index = raw.startIndex
        while index < raw.endIndex {
            let next = raw.index(index, offsetBy: 2)
            guard let byte = UInt8(raw[index..<next], radix: 16) else {
                throw SyncError.crypto("invalid hex")
            }
            data.append(byte)
            index = next
        }
        return data
    }

    /// HKDF-SHA256. The contract (shared/contract/SPEC.md §4) pins empty salt
    /// for all PICL derivations, which RFC 5869 §2.2 expands to `HashLen`
    /// zero bytes; the explicit parameter exists so the RFC 5869 known-answer
    /// fixture (`crypto-hkdf-rfc5869-case1`) can drive a non-empty salt too.
    static func hkdf(secret: Data, salt: Data = Data(), info: Data, length: Int) -> Data {
        let derived = HKDF<SHA256>.deriveKey(
            inputKeyMaterial: SymmetricKey(data: secret),
            salt: salt,
            info: info,
            outputByteCount: length
        )
        return derived.withUnsafeBytes { Data($0) }
    }

    static func namespacedHKDF(secret: Data, name: String, length: Int) -> Data {
        hkdf(secret: secret, info: Data((namespace + name).utf8), length: length)
    }

    static func xor(_ a: Data, _ b: Data) throws -> Data {
        guard a.count == b.count else { throw SyncError.crypto("xor length mismatch") }
        return Data(zip(a, b).map { $0 ^ $1 })
    }

    static func hmacSHA256(key: Data, data: Data) -> Data {
        Data(HMAC<SHA256>.authenticationCode(for: data, using: SymmetricKey(data: key)))
    }

    /// Constant-time comparison: XOR-accumulates every byte of the longer
    /// input with no early exit, so MAC verification never leaks through
    /// timing. True only when the lengths match AND the accumulated
    /// difference is zero.
    static func constantTimeEquals(_ a: Data, _ b: Data) -> Bool {
        var diff: UInt8 = 0
        for index in 0..<max(a.count, b.count) {
            let left = index < a.count ? a[a.startIndex + index] : 0
            let right = index < b.count ? b[b.startIndex + index] : 0
            diff |= left ^ right
        }
        return a.count == b.count && diff == 0
    }

    static func unbundle(bundleKey: Data, namespace name: String, payload: Data) throws -> Data {
        guard payload.count >= 32 else { throw SyncError.crypto("bundle too short") }
        let ciphertext = payload.dropLast(32)
        let expected = payload.suffix(32)
        let material = namespacedHKDF(secret: bundleKey, name: name, length: 32 + ciphertext.count)
        let hmacKey = material.prefix(32)
        let xorKey = material.dropFirst(32)
        let actual = hmacSHA256(key: Data(hmacKey), data: Data(ciphertext))
        guard constantTimeEquals(actual, expected) else { throw SyncError.crypto("bundle hmac mismatch") }
        return try xor(Data(ciphertext), Data(xorKey))
    }

    static func clientStateBytes(kB: Data) -> Data {
        Data(SHA256.hash(data: kB).prefix(16))
    }

    static func tokenMaterial(token: Data, type: String) -> (id: String, authKey: Data, bundleKey: Data) {
        let material = namespacedHKDF(secret: token, name: type, length: 96)
        return (
            hex(material.prefix(32)),
            Data(material.dropFirst(32).prefix(32)),
            Data(material.dropFirst(64).prefix(32))
        )
    }
}

enum SyncError: LocalizedError {
    case crypto(String)
    case auth(String)
    case network(String)
    case notSignedIn
    case totpRequired
    case storageUnavailable
    /// Conditional write conflict that survived the single retry mandated by
    /// SPEC §7.2, or a partial batch write.
    case conflict

    var errorDescription: String? {
        switch self {
        case .crypto(let m), .auth(let m), .network(let m): m
        case .notSignedIn: String(localized: "error.not_signed_in")
        case .totpRequired: String(localized: "error.totp")
        case .storageUnavailable: String(localized: "error.storage_unavailable")
        case .conflict: String(localized: "error.conflict")
        }
    }
}

File: android/app/src/main/java/de/kjell/zencompanion/sync/FxACrypto.kt
package de.kjell.zencompanion.sync

import java.security.MessageDigest
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

/** Ports of `Shared/SyncError` cases; UI maps these to friendly strings. */
sealed class SyncError(message: String?) : Exception(message) {
    class Crypto(detail: String? = null) : SyncError(detail)
    class Auth(detail: String? = null) : SyncError(detail)
    class Network(detail: String? = null) : SyncError(detail)

    /** Conditional write lost its race twice (SPEC §7.2); caller re-reads. */
    class Conflict(detail: String? = null) : SyncError(detail)
    object NotSignedIn : SyncError(null)
    object TotpRequired : SyncError(null)
    object StorageUnavailable : SyncError(null)

    companion object {
        const val CRYPTO_INVALID_HEX = "invalid hex"
    }
}

/** Ports of `Shared/FxACrypto.swift` — byte-for-byte identical outputs. */
object FxACrypto {
    const val NAMESPACE = "identity.mozilla.com/picl/v1/"

    fun hex(data: ByteArray): String = data.joinToString("") { "%02x".format(it) }

    fun unhex(string: String): ByteArray {
        val raw = string.trim()
        if (raw.length % 2 != 0) throw SyncError.Crypto(SyncError.CRYPTO_INVALID_HEX)
        val out = ByteArray(raw.length / 2)
        for (i in out.indices) {
            val hi = Character.digit(raw[i * 2], 16)
            val lo = Character.digit(raw[i * 2 + 1], 16)
            if (hi < 0 || lo < 0) throw SyncError.Crypto(SyncError.CRYPTO_INVALID_HEX)
            out[i] = ((hi shl 4) or lo).toByte()
        }
        return out
    }

    fun hmacSHA256(key: ByteArray, data: ByteArray): ByteArray {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(key, "HmacSHA256"))
        return mac.doFinal(data)
    }

    fun sha256(data: ByteArray): ByteArray =
        MessageDigest.getInstance("SHA-256").digest(data)

    /**
     * HKDF-SHA256 per RFC 5869. The app always calls this with the default
     * empty salt, which behaves exactly like CryptoKit's `Data()` salt
     * (HMAC zero-pads short keys); tests additionally pass RFC vectors.
     */
    fun hkdf(secret: ByteArray, info: ByteArray, length: Int, salt: ByteArray = ByteArray(32)): ByteArray {
        val effectiveSalt = if (salt.isEmpty()) ByteArray(32) else salt
        val prk = hmacSHA256(effectiveSalt, secret)
        val okm = ByteArray(length)
        var t = ByteArray(0)
        var counter = 1
        var offset = 0
        while (offset < length) {
            t = hmacSHA256(prk, t + info + byteArrayOf(counter.toByte()))
            val n = minOf(t.size, length - offset)
            System.arraycopy(t, 0, okm, offset, n)
            offset += n
            counter++
        }
        return okm
    }

    fun namespacedHKDF(secret: ByteArray, name: String, length: Int): ByteArray =
        hkdf(secret, (NAMESPACE + name).toByteArray(Charsets.UTF_8), length)

    fun xor(a: ByteArray, b: ByteArray): ByteArray {
        if (a.size != b.size) throw SyncError.Crypto("xor length mismatch")
        return ByteArray(a.size) { (a[it].toInt() xor b[it].toInt()).toByte() }
    }

    /** HKDF-unbundle: derive [hmacKey | xorKey], verify trailing HMAC, XOR-decrypt. */
    fun unbundle(bundleKey: ByteArray, namespaceName: String, payload: ByteArray): ByteArray {
        if (payload.size < 32) throw SyncError.Crypto("bundle too short")
        val ciphertext = payload.copyOfRange(0, payload.size - 32)
        val expected = payload.copyOfRange(payload.size - 32, payload.size)
        val material = namespacedHKDF(bundleKey, namespaceName, 32 + ciphertext.size)
        val hmacKey = material.copyOfRange(0, 32)
        val xorKey = material.copyOfRange(32, material.size)
        val actual = hmacSHA256(hmacKey, ciphertext)
        // Constant-time comparison so the tag check cannot leak timing.
        if (!MessageDigest.isEqual(actual, expected)) throw SyncError.Crypto("bundle hmac mismatch")
        return xor(ciphertext, xorKey)
    }

    fun clientStateBytes(kB: ByteArray): ByteArray = sha256(kB).copyOfRange(0, 16)

    /** Returns (id as hex, authKey, bundleKey). */
    fun tokenMaterial(token: ByteArray, type: String): Triple<String, ByteArray, ByteArray> {
        val material = namespacedHKDF(token, type, 96)
        return Triple(
            hex(material.copyOfRange(0, 32)),
            material.copyOfRange(32, 64),
            material.copyOfRange(64, 96),
        )
    }
}
